### PR TITLE
[SHELL32] Implement PathResolveA function

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose
Follow-up of #3762.
JIRA issue: [CORE-12665](https://jira.reactos.org/browse/CORE-12665)

## Proposed changes

- Add `PathResolveA` implementation by ANSI/Unicode conversion.

### CODE REVIEW

**Comment by katahiromz** [comment]
Check time. I will merge 3 days later.

**Comment by HBelusca** [comment]
Now I wonder: Does Windows 2003 really have a PathResolveA implemented function?

File: dll/win32/shell32/wine/shellpath.c [hunk 1]
```diff
@@ -631,8 +631,47 @@ VOID WINAPI PathQualifyAW(LPVOID pszPath)
...
+            if (dirsW[iDir] == NULL)
+                goto Cleanup;
+        }
+    }
+
+    SHAnsiToUnicode(path, pathW, _countof(pathW));
+
+    ret = PathResolveW(pathW, (LPCWSTR*)dirsW, flags);
+    if (ret)
+        SHUnicodeToAnsi(pathW, path, MAX_PATH);
```
Comment (HBelusca) [inline dll/win32/shell32/wine/shellpath.c:664]:
```suggestion
        SHUnicodeToAnsi(pathW, path, _countof(path));
```

File: dll/win32/shell32/wine/shellpath.c [hunk 2]
```diff
@@ -631,8 +631,47 @@ VOID WINAPI PathQualifyAW(LPVOID pszPath)
 
 BOOL WINAPI PathResolveA(LPSTR path, LPCSTR *dirs, DWORD flags)
 {
-    FIXME("(%s,%p,0x%08x),stub!\n", debugstr_a(path), dirs, flags);
-    return FALSE;
+    BOOL ret = FALSE;
+    LPWSTR *dirsW = NULL;
+    DWORD iDir, cDirs, cbDirs;
+    WCHAR pathW[MAX_PATH], tempW[MAX_PATH];
+    TRACE("PathResolveA(%s,%p,0x%08x)\n", debugstr_a(path), dirs, flags);
```
Comment (HBelusca) [inline dll/win32/shell32/wine/shellpath.c:638]:
```suggestion

    TRACE("PathResolveA(%s,%p,0x%08x)\n", debugstr_a(path), dirs, flags);
```

**Comment by katahiromz** [comment]
I cannot confirm. Win2k3's `PathResolve` doesn't call `PathResolveW`. Win2k3's `PathResolve` is the same as its `PathResolveW` (directly Implemented).

File: dll/win32/shell32/wine/shellpath.c (hunk 1 above)
Comment (katahiromz) [reply dll/win32/shell32/wine/shellpath.c:664]: `path` is a pointer.

**Comment by HBelusca** [comment]
> I cannot confirm the existence. Win2k3's `PathResolve` doesn't call `PathResolveW`. Win2k3's `PathResolve` is the same as its `PathResolveW` (directly implemented).

OK, and this is understandable.
Because when you look at this: https://git.reactos.org/?p=reactos.git;a=blob;f=dll/win32/shell32/wine/shellpath.c;hb=ada609bdbc18f1cb3b9bff1c6854c40be076b570#l718
that PathResolveA is called only when `SHELL_OsIsUnicode()` returns FALSE, which means, that for all NT OSes this doesn't execute.
So my opinion on this is: This is useless Wine code that wants to cover even the cases of the Win9x OSes...

File: dll/win32/shell32/wine/shellpath.c [hunk 3]
```diff
@@ -631,8 +631,48 @@ VOID WINAPI PathQualifyAW(LPVOID pszPath)
...
+        cbDirs = (cDirs + 1) * sizeof(LPWSTR);
+        dirsW = CoTaskMemAlloc(cbDirs);
+        if (!dirsW)
+            goto Cleanup;
+
+        ZeroMemory(dirsW, cbDirs);
+        for (iDir = 0; iDir < cDirs; ++iDir)
+        {
+            SHAnsiToUnicode(dirs[iDir], tempW, _countof(tempW));
+            SHStrDupW(tempW, &dirsW[iDir]);
```
Comment (HBelusca) [inline dll/win32/shell32/wine/shellpath.c:655]:
```suggestion
            dirsW[iDir] = __SHCloneStrAtoW(dirs[iDir], tempW);
```

**Comment by katahiromz** [comment]
Ok, I will close.

File: dll/win32/shell32/wine/shellpath.c [hunk 4]
```diff
@@ -631,8 +631,48 @@ VOID WINAPI PathQualifyAW(LPVOID pszPath)
...
+
+    ret = PathResolveW(pathW, (LPCWSTR*)dirsW, flags);
+    if (ret)
+        SHUnicodeToAnsi(pathW, path, MAX_PATH);
+
+Cleanup:
+    if (dirsW)
+    {
+        for (iDir = 0; iDir < cDirs; ++iDir)
+            CoTaskMemFree(dirsW[iDir]);
```
Comment (HBelusca) [inline dll/win32/shell32/wine/shellpath.c:671]:
```suggestion
            SHFree(dirsW[iDir]);
```

File: dll/win32/shell32/wine/shellpath.c [hunk 5]
```diff
@@ -631,8 +631,48 @@ VOID WINAPI PathQualifyAW(LPVOID pszPath)
...
+
+    TRACE("PathResolveA(%s,%p,0x%08x)\n", debugstr_a(path), dirs, flags);
+
+    if (dirs)
+    {
+        for (cDirs = 0; dirs[cDirs]; ++cDirs)
+            ;
+
+        cbDirs = (cDirs + 1) * sizeof(LPWSTR);
+        dirsW = CoTaskMemAlloc(cbDirs);
```
Comment (HBelusca) [inline dll/win32/shell32/wine/shellpath.c:647]:
```suggestion
        dirsW = SHAlloc(cbDirs);
```

File: dll/win32/shell32/wine/shellpath.c [hunk 6]
```diff
@@ -631,8 +631,48 @@ VOID WINAPI PathQualifyAW(LPVOID pszPath)
...
+    if (ret)
+        SHUnicodeToAnsi(pathW, path, MAX_PATH);
+
+Cleanup:
+    if (dirsW)
+    {
+        for (iDir = 0; iDir < cDirs; ++iDir)
+            CoTaskMemFree(dirsW[iDir]);
+
+        CoTaskMemFree(dirsW);
```
Comment (HBelusca) [inline dll/win32/shell32/wine/shellpath.c:673]:
```suggestion
        SHAlloc(dirsW);
```

File: dll/win32/shell32/wine/shellpath.c (hunk 3 above)
Comment (katahiromz) [reply dll/win32/shell32/wine/shellpath.c]: OK.

File: dll/win32/shell32/wine/shellpath.c (hunk 5 above)
Comment (katahiromz) [reply dll/win32/shell32/wine/shellpath.c]: OK.

File: dll/win32/shell32/wine/shellpath.c (hunk 6 above)
Comment (katahiromz) [reply dll/win32/shell32/wine/shellpath.c]: `SHFree` is used.

File: dll/win32/shell32/wine/shellpath.c (hunk 2 above)
Comment (katahiromz) [reply dll/win32/shell32/wine/shellpath.c:639]: OK.

File: dll/win32/shell32/wine/shellpath.c (hunk 4 above)
Comment (katahiromz) [reply dll/win32/shell32/wine/shellpath.c]: OK.

**Comment by katahiromz** [comment]
Check time. I will merge 3 days later.

**Comment by katahiromz** [comment]
To go?

**Pull request merged**